### PR TITLE
[HTTP2] More Integration Tests

### PR DESCRIPTION
**Pull request by dnadoba**

### CODE REVIEW

File: Tests/AsyncHTTPClientTests/HTTP2ClientTests.swift [hunk 1]
```diff
@@ -187,4 +201,169 @@ class HTTP2ClientTests: XCTestCase {
...
+        task = client.execute(
+            request: request,
+            delegate: delegate
+        )
+
+        XCTAssertThrowsError(try task.futureResult.timeout(after: .seconds(2)).wait(), "Should fail") { error in
+            guard case let error = error as? HTTPClientError, error == .cancelled else {
+                return XCTFail("Should fail with cancelled")
+            }
+        }
```
Comment (fabianfett) [inline Tests/AsyncHTTPClientTests/HTTP2ClientTests.swift:267]:
```suggestion
        XCTAssertThrowsError(try task.futureResult.timeout(after: .seconds(2)).wait()) {
            XCTAssertEqual($0 as? HTTPClientError, . cancelled)
        }
```

File: Tests/AsyncHTTPClientTests/HTTP2ClientTests.swift [hunk 2]
```diff
@@ -187,4 +201,169 @@ class HTTP2ClientTests: XCTestCase {
...
+        for result in results {
+            switch result {
+            case .success:
+                XCTFail("Shouldn't succeed")
+            case .failure(let error):
+                if let clientError = error as? HTTPClientError, clientError == .cancelled {
+                    continue
+                } else {
+                    XCTFail("Unexpected error: \(error)")
+                }
```
Comment (fabianfett) [inline Tests/AsyncHTTPClientTests/HTTP2ClientTests.swift:236]:
```suggestion
                XCTAssert(error as? HTTPClientError, .cancelled)
```

File: Tests/AsyncHTTPClientTests/HTTP2ClientTests.swift [hunk 3]
```diff
@@ -187,4 +201,169 @@ class HTTP2ClientTests: XCTestCase {
...
+            )
+            return task
+        }
+
+        for task in tasks {
+            XCTAssertThrowsError(try task.futureResult.timeout(after: .seconds(2)).wait(), "Should fail") { error in
+                guard case let error = error as? HTTPClientError, error == .cancelled else {
+                    return XCTFail("Should fail with cancelled")
+                }
+            }
```
Comment (fabianfett) [inline Tests/AsyncHTTPClientTests/HTTP2ClientTests.swift:305]:
```suggestion
            XCTAssertThrowsError(try task.futureResult.timeout(after: .seconds(2)).wait()) {
                XCTAssertEqual($0 as? HTTPClientError, .cancelled)
            }
```

File: Tests/AsyncHTTPClientTests/HTTPClientTestUtils.swift [hunk 4]
```diff
@@ -804,6 +813,10 @@ internal final class HTTPBinHandler: ChannelInboundHandler {
                 builder.add(buff)
                 self.resps.append(builder)
                 return
+            case "/sendheaderandwait":
+                // sends some headers and waits indefinitely afterwards
+                context.writeAndFlush(wrapOutboundOut(.head(HTTPResponseHead(version: HTTPVersion(major: 1, minor: 1), status: .ok))), promise: nil)
+                return
```
Comment (fabianfett) [inline Tests/AsyncHTTPClientTests/HTTPClientTestUtils.swift:819]: Please create a small `ChannelInboundHandler` for this in the specific test case. There is so much stuff in this single handler and all is mixed.

File: Tests/AsyncHTTPClientTests/HTTPClientTestUtils.swift [hunk 5]
```diff
@@ -49,11 +49,16 @@ func getDefaultEventLoopGroup(numberOfThreads: Int) -> EventLoopGroup {
...
+    init(
+        backpressureEventLoop: EventLoop? = nil,
+        stateDidChangeCallback: ((State) -> Void)? = nil
+    ) {
         self.backpressureEventLoop = backpressureEventLoop
+        self.stateDidChangeCallback = stateDidChangeCallback
     }
 
     var backpressureEventLoop: EventLoop?
+    var stateDidChangeCallback: ((State) -> Void)?
```
Comment (fabianfett) [inline Tests/AsyncHTTPClientTests/HTTPClientTestUtils.swift:61]: this should be at least a let. however in your tests your mainly using this to get a callback once the head was received. I think, a specialized delegate only for this one job, might be a better idea.

**Review by fabianfett — CHANGES_REQUESTED** [review]

File: Tests/AsyncHTTPClientTests/HTTP2ClientTests.swift [hunk 6]
```diff
@@ -67,7 +82,7 @@ class HTTP2ClientTests: XCTestCase {
     func testConcurrentRequestsFromDifferentThreads() {
         let bin = HTTPBin(.http2(compress: false))
         defer { XCTAssertNoThrow(try bin.shutdown()) }
-        let client = self.makeDefaultHTTPClient()
+        let client = self.makeClientWithActiveHTTP2Connection(to: bin)
```
Comment (fabianfett) [inline Tests/AsyncHTTPClientTests/HTTP2ClientTests.swift:85]: why do we need a client with an active connection?

Comment (dnadoba) [reply Tests/AsyncHTTPClientTests/HTTP2ClientTests.swift]: good catch, we don't.

File: Tests/AsyncHTTPClientTests/HTTP2ClientTests.swift [hunk 7]
```diff
@@ -187,4 +202,185 @@ class HTTP2ClientTests: XCTestCase {
         // all workers should be running, let's wait for them to finish
         allDone.wait()
     }
+
+    func testUncleanShutdownCancelsExecutingAndQueuedTasks() {
+        let bin = HTTPBin(.http2(compress: false))
+        defer { XCTAssertNoThrow(try bin.shutdown()) }
+        let clientGroup = MultiThreadedEventLoopGroup(numberOfThreads: 1)
+        defer { XCTAssertNoThrow(try clientGroup.syncShutdownGracefully()) }
+        let client = self.makeClientWithActiveHTTP2Connection(to: bin, eventLoopGroupProvider: .shared(clientGroup))
```
Comment (fabianfett) [inline Tests/AsyncHTTPClientTests/HTTP2ClientTests.swift:211]: Why do we need a client with an active connection here? Would love a comment...

File: Tests/AsyncHTTPClientTests/HTTP2ClientTests.swift [hunk 8]
```diff
@@ -187,4 +202,185 @@ class HTTP2ClientTests: XCTestCase {
...
+private final class SendHeaderAndWaitChannelHandler: ChannelInboundHandler {
+    typealias InboundIn = HTTPServerRequestPart
+    typealias OutboundOut = HTTPServerResponsePart
+
+    func channelRead(context: ChannelHandlerContext, data: NIOAny) {
+        context.writeAndFlush(wrapOutboundOut(.head(HTTPResponseHead(
+            version: HTTPVersion(major: 1, minor: 1),
+            status: .ok
+        ))
+        ), promise: nil)
```
Comment (fabianfett) [inline Tests/AsyncHTTPClientTests/HTTP2ClientTests.swift:384]: This will send at least two response headers for each request... you'll need to unwrap the incoming requests for `.head` and `.end`

File: Tests/AsyncHTTPClientTests/HTTPClientTestUtils.swift [hunk 9]
```diff
@@ -49,11 +49,16 @@ func getDefaultEventLoopGroup(numberOfThreads: Int) -> EventLoopGroup {
 class TestHTTPDelegate: HTTPClientResponseDelegate {
     typealias Response = Void
 
-    init(backpressureEventLoop: EventLoop? = nil) {
+    init(
+        backpressureEventLoop: EventLoop? = nil,
+        stateDidChangeCallback: ((State) -> Void)? = nil
+    ) {
```
Comment (fabianfett) [inline Tests/AsyncHTTPClientTests/HTTPClientTestUtils.swift:55]: I think all changes in this file can now be reverted, can't they?

**Review by fabianfett — COMMENTED** [review]
Three small things at the end.

File: Tests/AsyncHTTPClientTests/HTTP2ClientTests.swift [hunk 10]
```diff
@@ -187,4 +202,185 @@ class HTTP2ClientTests: XCTestCase {
...
+    func didReceiveHead(task: HTTPClient.Task<Void>, _ head: HTTPResponseHead) -> EventLoopFuture<Void> {
+        self.didReceiveHeadCallback(head)
+        return task.eventLoop.makeSucceededVoidFuture()
+    }
+
+    func didFinishRequest(task: HTTPClient.Task<Void>) throws {}
+}
+
+/// sends some headers and waits indefinitely afterwards
+private final class SendHeaderAndWaitChannelHandler: ChannelInboundHandler {
```
Comment (fabianfett) [inline Tests/AsyncHTTPClientTests/HTTP2ClientTests.swift:375]: oh we should use this one to test idleReadTimeout as well. can be another pr.

File: Tests/AsyncHTTPClientTests/HTTP2ClientTests.swift (hunk 7 above)
Comment (dnadoba) [reply Tests/AsyncHTTPClientTests/HTTP2ClientTests.swift:211]: We need an active connection to guarantee that requests are executed immediately without waiting for connection establishment. Maybe we could also write a test that observes when a connection has been established and only starts shutdown afterwards.

File: Tests/AsyncHTTPClientTests/HTTP2ClientTests.swift (hunk 8 above)
Comment (dnadoba) [reply Tests/AsyncHTTPClientTests/HTTP2ClientTests.swift:384]: could we also use `channelActive` instead?

File: Tests/AsyncHTTPClientTests/HTTP2ClientTests.swift (hunk 7 above)
Comment (fabianfett) [reply Tests/AsyncHTTPClientTests/HTTP2ClientTests.swift:213]: would you mind putting this into the test?

File: Tests/AsyncHTTPClientTests/HTTP2ClientTests.swift (hunk 8 above)
Comment (fabianfett) [reply Tests/AsyncHTTPClientTests/HTTP2ClientTests.swift:386]: nope... you need to wait for the request.

File: Tests/AsyncHTTPClientTests/HTTP2ClientTests.swift (hunk 7 above)
Comment (dnadoba) [reply Tests/AsyncHTTPClientTests/HTTP2ClientTests.swift:213]: Done

File: Tests/AsyncHTTPClientTests/HTTP2ClientTests.swift (hunk 8 above)
Comment (dnadoba) [reply Tests/AsyncHTTPClientTests/HTTP2ClientTests.swift]: okay thanks, we now only send it after we received `.end`.

**Review by fabianfett — APPROVED** [review]
